### PR TITLE
Upload imported bundles over HTTP instead of the WebSocket

### LINKED ISSUE #2327: Make UI bundle import work for large bundles via HTTP upload (WS message cap blocks it today)
### Summary

The UI "import bundle" upload (`devices/import_bundle`) sends the entire bundle as a single base64 field inside one WebSocket command message, so it's hard-capped by the dashboard `/ws` `max_msg_size` — **not** by `BUNDLE_MAX_TOTAL_BYTES`. This is the browser → dashboard path, distinct from the offloader → receiver peer-link upload (#2326, closed as won't-do).

### The limit

- `import_bundle` is an `@api_command` taking the full `file_content_b64: str` in one call (`controllers/devices/controller.py`), so the whole payload rides one `/ws` message.
- `ws.py` creates `web.WebSocketResponse(heartbeat=...)` with **no `max_msg_size`**, so it uses aiohttp's default of **4 MiB** (4194304).
- base64 inflates ~4/3 and the JSON envelope adds overhead, so a raw bundle over **~3 MiB** already exceeds the frame and the WS drops it — likely as an opaque connection error, not a clean "bundle too large".
- Consequence: the `_MAX_BUNDLE_UPLOAD_BYTES` guard (now `BUNDLE_MAX_TOTAL_BYTES`, 128 MiB after #2324) never binds on this path — the WS caps first. The cap value is effectively cosmetic for the UI, and the #2324 framing ("a config that can be remote-built can also be imported") is inverted here: the UI import is far *more* restrictive (~3 MiB) than remote build (128 MiB over chunked peer-link frames).

Image/font-heavy configs — the exact case behind the original #17837 / #2320 work — are the ones that produce bundles large enough to hit this, so a user exporting and re-importing such a device via the UI can't.

### Direction: HTTP upload

Make #2324's 128 MiB cap actually reachable by moving the UI bundle upload to an **HTTP `POST`/`PUT` endpoint** instead of a base64 WS command, streamed to disk. This mirrors the existing `GET /api/firmware/download` (`controllers/firmware/download.py`), which was moved to HTTP precisely because a large `firmware.elf` "isn't capped by a proxy's WebSocket `max_msg_size`". Sketch:

- Add an authed HTTP upload route (reuse the download side's single-use capability-token pattern — `firmware/download_token` → token consumed by the route — so the upload is authorized without a base64 WS round-trip).
- Stream the request body straight to the staging tarball, then run the existing `_stage_bundle` extract/plan/place logic on it (the `import_bundle` two-phase conflict flow stays; only the transport for the bytes changes).
- Keep `set client_max_size` on the route (aiohttp `Application`/route body limit) aligned with `BUNDLE_MAX_TOTAL_BYTES` so the cap #2324 set is the real, enforced limit — with a clean "bundle too large" error instead of an opaque WS drop.
- Companion frontend change: `POST` the file instead of sending `file_content_b64` over the WS command.

The WS command can stay for tiny bundles / backward compat, or be retired once the frontend moves over.

### Frontend coordination

The frontend (`esphome/device-builder-frontend`) sends `import_bundle`; options 1 and 3 need a companion frontend PR.

### Related

- #2324 (raised the import cap to the shared 128 MiB — never binds on this WS path)
- #2326 (peer-link upload re-arch, closed as won't-do — different path)
- Precedent: `GET /api/firmware/download` moved off WS for the same `max_msg_size` reason

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

Move the "import bundle" upload from the `devices/import_bundle` WebSocket command (which base64-encoded the whole `.tar.gz` into one WS message, capped at aiohttp's 4 MiB `/ws` `max_msg_size`) to the new `POST /api/devices/import_bundle` HTTP route, so an image/font-heavy bundle up to the shared 128 MiB limit can be imported through the UI.

`importBundleUpload(file, overwrite?)` mints a single-use token over the WS (`devices/import_bundle_token`) and `POST`s the `File` straight to the route, mirroring how `firmwareDownloadUrl` authorizes the HTTP download route with a token in the query string (no `Authorization` header). It parses the same `ImportBundleResponse` shape the WS command returned, so the two-phase conflict flow, the resolve-conflicts step, and the partial-import step are unchanged. The import-flow controller now re-uploads the retained `File` on resolve instead of caching base64, so the base64 read is gone.

Requires the backend companion (adds the route + token command and removes the old WS command); they land in lock-step.

**Related issue or feature (if applicable):**

- fixes https://github.com/esphome/device-builder/issues/2327
- backend companion: https://github.com/esphome/device-builder/pull/2340

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue) — `bugfix`
- [ ] New feature (non-breaking change which adds functionality) — `new-feature`
- [x] Enhancement to an existing feature — `enhancement`
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected) — `breaking-change`
- [ ] Refactor (no behaviour change) — `refactor`
- [ ] Documentation only — `docs`
- [ ] Maintenance / chore — `maintenance`
- [ ] CI / workflow change — `ci`
- [ ] Dependencies bump — `dependencies`

## Checklist

- [x] The code change is tested and works locally.
- [x] `pnpm run lint` passes.
- [x] `pnpm run test` passes.
- [x] Tests have been added to verify that the new code works (where applicable).

### CODE REVIEW

**Comment by bdraco** [comment]
@esphbot review

**Comment by bdraco** [comment]
@esphbot review

File: src/components/wizard/import-flow-controller.ts [hunk 1]
```diff
@@ -157,31 +153,19 @@ export class ImportFlowController implements ReactiveController {
...
-      }
-
-      const res = await this._host.api.importBundle({
-        file_content_b64: this._bundleB64,
-        ...(overwrite !== undefined ? { overwrite } : {}),
-      });
+      const res = await this._host.api.importBundleUpload(this._file, overwrite);
       if (res.status === "conflicts") {
         this.conflicts = res.conflicts;
         this.hasSecrets = res.has_secrets;
```
[the remark this answers is not in the document]
Comment (bdraco) [reply src/components/wizard/import-flow-controller.ts:183]: Good catch, fixed in 53f3556 — start() now bails when importBusy, so a file picked while an import is in flight can't overwrite this._file before the conflict round-trip re-uploads it. Added a regression test covering it.

File: src/api/esphome-api.ts [hunk 2]
```diff
@@ -806,17 +806,45 @@ export class ESPHomeAPI {
...
+      try {
+        const body = await response.json();
+        errorCode = body.error_code ?? errorCode;
+        details = body.details ?? details;
+      } catch {
+        // A non-JSON body (e.g. a 413's plain text) keeps the generic message.
+      }
+      throw new APIError(errorCode, details);
+    }
+    return response.json() as Promise<ImportBundleResponse>;
```
[the remark this answers is not in the document]
Comment (bdraco) [reply src/api/esphome-api.ts]: Added in 53f3556 — a DRY 'import bundle upload (HTTP)' block in test/api/esphome-api.test.ts mirroring the firmware-download tests: token minting over the WS, base-path/token URL construction, mode=resolve + repeated overwrite params, and both the JSON {error_code, details} and non-JSON (413) error-parsing paths.

File: src/api/esphome-api.ts [hunk 3]
```diff
@@ -806,17 +806,43 @@ export class ESPHomeAPI {
...
+    if (!response.ok) {
+      let errorCode = "http_error";
+      let details = `Bundle upload failed: ${response.status}`;
+      try {
+        const body = await response.json();
+        errorCode = body.error_code ?? errorCode;
+        details = body.details ?? details;
+      } catch {
+        // A non-JSON body (e.g. a 413's plain text) keeps the generic message.
+      }
```
[the remark this answers is not in the document]
Comment (bdraco) [reply src/api/esphome-api.ts:846]: Fixed in 53f3556 — the parse now narrows the body to an object and type-checks error_code/details before reading them (so a null/odd shape can't throw), and the fallback message now includes statusText.

**Comment by bdraco** [comment]
@esphbot review

File: src/components/wizard/import-flow-controller.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (esphbot) [reply src/components/wizard/import-flow-controller.ts:183]: Verified. Guard holds for a real reason: `_importBundleFlow` sets `importBusy = true` synchronously, before its first `await`. Second same-tick pick sees it set. Ordering survives the `void this._createImportedDevice()` hop. Dropped file stays visible to the user — `importBusy` maps to `_submitting`, which drives `?busy` on `esphome-base-dialog`.

File: src/api/esphome-api.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (esphbot) [reply src/api/esphome-api.ts]: Confirmed. Block mirrors `firmware download (HTTP)` exactly — same install/teardown, same `sentAs(0)`. One branch still uncovered: `overwrite: []`. Reachable by default — `_overwrite` starts empty, Import button ungated. Emits `mode=resolve` with zero `overwrite` params. Left a suggestion.

File: src/api/esphome-api.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (esphbot) [reply src/api/esphome-api.ts:846]: Confirmed. `body && typeof body === "object"` rejects `null` pre-destructure. Per-field `typeof` checks stop a hostile shape throwing inside the handler. Arrays fall through harmlessly to `undefined`. `statusText` now in the fallback.

**Review by esphbot — APPROVED** [review]
> [!TIP]
> No blocking issues found — ready to merge.

**Comment by bdraco** [comment]
@esphbot review

File: src/api/esphome-api.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (bdraco) [reply src/api/esphome-api.ts]: Covered in b5044a8 — added a test asserting an empty resolve emits `?mode=resolve` with no `overwrite` params.

**Comment by bdraco** [comment]
@esphbot review

File: src/components/wizard/import-flow-controller.ts [hunk 4]
```diff
@@ -71,6 +67,10 @@ export class ImportFlowController implements ReactiveController {
 
   /** Begin importing the picked file (YAML upload or bundle). */
   start(file: File): void {
+    // Don't clobber the file/state of an import already in flight — its
+    // conflict round-trip re-uploads this._file, so overwriting it mid-flight
+    // would resolve against the wrong archive.
+    if (this._host.importBusy) return;
```
[the remark this answers is not in the document]
Comment (bdraco) [reply src/components/wizard/import-flow-controller.ts:73]: Fixed in 347cb29 — _createImportedDevice now rejects a pick that is neither a recognised bundle nor YAML (.yaml/.yml) with a clear 'unsupported file type' error, so the bare .gz the macOS picker admits can't be read as text. Added a regression test.

File: test/api/esphome-api.test.ts [hunk 5]
```diff
@@ -2191,6 +2191,113 @@ describe("ESPHomeAPI — firmware download (HTTP)", () => {
...
+    vi.stubGlobal("fetch", fetchFn);
+    const file = new File([new Uint8Array([1])], "device.esphomebundle.tar.gz");
+
+    await upload(api, ws, file, ["device.yaml", "common/wifi.yaml"]);
+
+    expect(fetchFn).toHaveBeenCalledWith(
+      "/api/devices/import_bundle?token=a+b%2Fc&mode=resolve&overwrite=device.yaml&overwrite=common%2Fwifi.yaml",
+      { method: "POST", body: file }
+    );
+  });
```
[the remark this answers is not in the document]
Comment (bdraco) [reply test/api/esphome-api.test.ts:2257]: Already covered — b5044a8 added a test asserting the empty resolve emits ?mode=resolve with zero overwrite params (distinct from the detect pass's no-mode URL).

**Comment by bdraco** [comment]
@esphbot review

**Comment by bdraco** [comment]
Thanks for the review. On the three suggestions:

1. **No upload progress / cancellation** — intentionally skipping for now. Most bundles are small, so the busy modal is short in practice, and the abort/`XMLHttpRequest`-progress work isn't worth it here. Can revisit if large bundles turn out to be common.
2. **Empty-resolve (`overwrite: []`) untested** — added in b5044a8: a test asserting the keep-all resolve emits `?mode=resolve` with zero `overwrite=` params, distinct from the detect pass's no-`mode` URL, so a future `if (overwrite?.length)` regression would fail the suite.
3. **Stale `ImportBundleResponse` docstring** — fixed in 347cb29: now "Response from `POST /api/devices/import_bundle`" and "re-uploads the same file", so all three docstrings agree.

Also hardened the file picker in 347cb29: the bare `.gz` added for macOS' compound-extension quirk is now rejected post-selection unless the file is a real bundle or YAML, so it can't route binary through the text-import path.

File: src/translations/en.json [hunk 6]
```diff
@@ -581,6 +581,7 @@
     "add_board": "Select",
     "import_read_error": "Failed to read the file.",
     "import_general_error": "Failed to import the config file. Please try again.",
+    "import_unsupported_type": "Unsupported file type. Import a YAML config (.yaml, .yml) or an ESPHome bundle (.tar.gz).",
```
[the remark this answers is not in the document]
Comment (bdraco) [reply src/translations/en.json]: Fixed in 568f83a — the message now lists all three bundle extensions (.tar.gz, .tgz, .esphomebundle) so a .tgz / .esphomebundle pick doesn't read as unsupported.

**Comment by bdraco** [comment]
@esphbot review

File: src/components/wizard/import-flow-controller.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (esphbot) [reply src/components/wizard/import-flow-controller.ts:73]: Verified. Gate sits after the `isBundleFilename` route, so only picker-widened files reach it. Drag-drop unaffected — `FileDropController` still defaults to `ACCEPTED_UPLOAD_EXTENSIONS` (`file-drop-controller.ts:40`), never `FILE_INPUT_ACCEPT`. Re-pick works too: `wizard-step-method.ts:249` clears `_fileInput.value` before dispatch. One nit: `isYamlFilename` itself lacks a direct unit test — `.yml` and case-insensitivity unpinned.

**Review by esphbot — APPROVED** [review]
> [!TIP]
> No blocking issues found — ready to merge.

**Comment by bdraco** [comment]
Addressed the one nit in aafe3c6 — added a direct `isYamlFilename` unit test pinning the `.yml` alias and case-insensitivity (`DEVICE.YML`), mirroring the `isBundleFilename` block.

**Comment by bdraco** [comment]
@esphbot review

**Review by esphbot — APPROVED** [review]
> [!TIP]
> No blocking issues found — ready to merge.
